tests/upgrade_tests/test_firmware_upgrades.py

```diff
@@ -194,6 +198,47 @@ def asserts(tag, client):
         assert btc.get_address(emu.client, "Bitcoin", PATH) == address
 
 
+# Although Shamir was introduced in 2.1.2 already, the debug instrumentation was not present until 2.1.9.
+@for_all("core", minimum_version=(2, 1, 9))
+def test_upgrade_shamir_recovery(gen, from_tag, to_tag):
+    with EmulatorWrapper(gen, from_tag) as emu:
+        assert emu.client.features.recovery_mode is False
+
+        device_handler = BackgroundDeviceHandler(emu.client)
```

Comment:
this is the hang problem. you need to use `device_handler` as a context manager, so, somewhat awkwardly:
`with EmulatorWrapper(...) as emu, BackgroundDeviceHandler(emu.client) as device_handler:`

Reply:
Oh, right! Will do

Reply:
at least I think. let me try at home

Reply:
..yes. whenever you're using `device_handler.run`, you need to ensure that the finalizer runs. the ctxmanager does that automatically.

Reply:
Yes, indeed. I confirm.